consensus/src/quorum_store/batch_generator.rs

```diff
@@ -160,36 +157,59 @@ impl BatchGenerator {
         }
     }
 
-    fn build_account_min_gas(
-        exclude_txns: &[TransactionInProgress],
-        pulled_txns: &[SignedTransaction],
-    ) -> HashMap<PeerId, u64> {
-        let mut account_min_gas = HashMap::new();
-        for txn in exclude_txns {
-            let current_min = account_min_gas
-                .entry(txn.summary.sender)
-                .or_insert(txn.gas_unit_price);
-            if txn.gas_unit_price < *current_min {
-                *current_min = txn.gas_unit_price
+    fn bucket_into_batches(
+        &mut self,
+        pulled_txns: &mut Vec<SignedTransaction>,
+        expiry_time: u64,
+    ) -> Vec<Batch> {
+        pulled_txns.sort_by_key(|txn| txn.gas_unit_price());
+
+        let mut batches = vec![];
+        let buckets = self.config.batch_buckets.clone();
+        for (&bucket_start, &bucket_end) in buckets.iter().tuple_windows() {
+            if pulled_txns.is_empty() {
+                break;
             }
-        }
-        for txn in pulled_txns {
-            let current_min = account_min_gas
-                .entry(txn.sender())
-                .or_insert_with(|| txn.gas_unit_price());
-            if txn.gas_unit_price() < *current_min {
-                *current_min = txn.gas_unit_price()
+
+            let num_txns_in_bucket = match pulled_txns
+                .binary_search_by_key(&(bucket_end, PeerId::ZERO), |txn| {
```

Comment:
Are you sure this is what you want? From the documentation: "If there are multiple matches, then any one of the matches could be returned". Meaning that txns with the same gas can go to different buckets, right?

Reply:
The key for the binary search includes the PeerId, and the search key includes PeerId::ZERO. So we don't have to worry about multiple matches. (AFAICT PeerId::ZERO is not a valid address; it's used as a placeholder for "null" in a few places in the code.)

Reply:
it means it's not supposed to return Ok(..) right?